common/devdoc/refcount_requirements.md

```diff
@@ -13,14 +13,25 @@ It wraps the structure that needs to be ref counted into another structure that
 #define REFCOUNT_TYPE_CREATE_WITH_EXTRA_SIZE(type, size) MU_C2(REFCOUNT_SHORT_TYPE(type), _Create_With_Extra_Size)(size)
 #define REFCOUNT_TYPE_DESTROY(type, var) MU_C2(REFCOUNT_SHORT_TYPE(type), _Destroy)(var)
 
-#define DEFINE_REFCOUNT_TYPE(type) \
+#define INC_REF(type, var) interlocked_increment(&((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
+#define DEC_REF(type, var) interlocked_decrement(&((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
+#define INIT_REF(type, var) interlocked_exchange(&((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count, 1)
+
+#define DEFINE_REFCOUNT_TYPE(type, malloc_func, free_func) \
 ...
 ```
 
 ### DEFINE_REFCOUNT_TYPE
 
+```c
+#define DEFINE_REFCOUNT_TYPE(type, malloc_func, free_func) \
```

Comment:
>DEFINE_REFCOUNT_TYPE [](start = 8, length = 20)

Will there be defaults for the malloc and free functions used? If so, Should we have a different macro `DEFINE_REFCOUNT_TYPE_CUSTOM_ALLOC` so that  this does not ripple through the repos?

Reply:
A very good idea. I do not remember why I discarded it initially. It sounds much better when you suggested it :-).

---
In reply to: [470943363](https://github.com/Azure/azure-c-pal/pull/33#discussion_r470943363) [](ancestors = 470943363)